### DISCUSSION TITLE
some notes about how ggml works using the GPT-2 example

### DISCUSSION THREAD

**Discussion by chunhualiao** [ggml-org/ggml#716] [Show and tell]
It might be useful for this library to have some documentation about how it works internally. 
I may have missed such documentation somewhere so I went ahead to debug the inference example using gpt-2 117M. 

Some initial notes are put here https://gist.github.com/chunhualiao/8610c8a3afa3ef76c0174c57ff6e5339
This may be useful for beginners  though it may contain errors of course. 

A snapshot of my notes is copied below:

I have taken quite some machine learning courses and have done a few projects already. I think I know the math formula involved in transformers and GPT models. However, I always wondered how they work in reality.  The best way for me is to read and understand source codes implementing these models.
I am a C/C++ programmer mostly. I am more comfortable to read C/C++ programs. So, recently I started to read, run, and debug ggml's gpt-2 inference example since ggml is entirely written in C and can run many transformer models on a laptop: https://github.com/ggerganov/ggml/tree/master/examples/gpt-2 . The famous llama.cpp is closely connected to this library. My experiment environment is a MacBook Pro laptop+ Visual Studio Code + cmake+ CodeLLDB (gdb does not work with my M2 chip), and GPT-2 117 M model.

Here is what I have learned so far:

The high-level main function has the following structure https://github.com/ggerganov/ggml/blob/master/examples/gpt-2/main-backend.cpp
- load the model: ggml specific format using quantization. 
- create a compute graph from the loaded model. I will explain this graph later.
- tokenized the prompt
- using a loop to feed the prompt into the model, and generate a new token each iteration
  - Inside the loop, the prompt is fed into the model's compute graph
  - when the compute graph is walked through entirely, the last node stores the results to help choose the next new token
  - generate a new token using  the top-K and top-P sampling algorithm
  - update the prompt to include the new token, the prompt will be used in the next iteration

The core computation is done using the compute graph.
- all computations involved in a neural network/model's inference can be modeled by using some input vector/matrix to compute a resulting vector/matrix.  
- If we focus on each vector and matrix, we can model the computing as forward walking/updating a directed graph: each node of the graph is a tensor, representing a vector or matrix
- Each node/tensor stores its value and pointers to relevant input nodes/tensors and operations.  The result is written back to the current tensor. 
- The inference now becomes the walk of the graph from the beginning to the end, following the edges from one tensor to another, updating each tensor's value based on the inputs and operations.

ggml provides quite some tools to dump or visualize the compute graph, which helps debug the inference process. https://netron.app/ also can visualize common model files hosted on huggingface. I tried to upload huggingface GPT-2 model to netron. It is fascinating to view the compute graph of a transformer model. 

ggml has many other advanced features including running computation on GPUs, using multi-threaded programming, and so on. 

Even for a small model like GPT-2 117M, the compute graph is quite large (leaf nodes 188 + non-leaf nodes 487). I will need more time to go through the graph to have a deeper understanding of how all the math formula of transformers is implemented in a programming language. 

I have tremendous respect for ggml/llama.cpp's author: Georgi Gerganov.   What a genius to pull off some projects like this!

**Comment by Nick-infinity** [comment]
I am also trying to understand the e2e ggml flow.  I am stuck at the memory alloc part. It seems like there are lot of diff buffers that needs to be allocated apart from tensor data buffer and programmers need to take care of overheads manually in GMML. 
I think the examples and test cases provides a very basic view. 
Thank you for sharing this.

**Reply by slaren** [reply]
For the most part, if you are using the ggml-backend API, you shouldn't need to worry too much about the overheads. You will need to consider the number of tensors in your model, but that should be relatively straightforward. I would suggest starting from the [simple-backend example](https://github.com/ggerganov/ggml/blob/master/examples/simple/simple-backend.cpp), and once you are comfortable with that you can move onto the [gpt2 backend example](https://github.com/ggerganov/ggml/blob/master/examples/gpt-2/main-backend.cpp) for an example of how to implement a transformer model with ggml. If you have any specific questions don't hesitate to ask.

**Reply by teleprint-me** [reply]
@slaren I'm trying to create a simple MLP model by starting out with the forward pass for the compute graph, but the input has garbage shapes for some reason. I'm probably making a silly mistake, but I'm having a hard time comprehending why this is the case. The simple-backend example sets the inputs before the model is "loaded" and then it's initialized (according to my current understanding).

```sh
model->x: ne=[8 1 1 1], nb=[4 32 32 32]
out: ne=[103314503939032 0 0 103314503955416], nb=[103314503971800 4099 103314504020992 103314503988200]
layer->W: ne=[8 16 1 1], nb=[4 32 512 512]
layer->b: ne=[16 1 1 1], nb=[4 64 64 64]
```

It's not obvious to me why I need to know the input in advance. I thought maybe I just needed to know the shapes and the amount of memory needed for reservation, but I'm admittedly stuck on this.

Model code can be found here: https://github.com/teleprint-me/bugbear/blob/main/examples/mlp.cpp

I understand you all are very busy. Any feedback is appreciated.

**Reply by slaren** [reply]
It looks like the tensor struct is being overwritten with garbage data. This can happen if you reuse the same buffer for different `ggml_context` objects, but it may also be caused by other memory issues. Your code adds some abstractions that make it harder to see exactly what's going on. I suggest that you start with very simple code, and don't try to abstract the ggml API into classes until you are confident that you understand how it works.

If you have any specific questions about ggml feel free to ask, but try to keep it focused on a single point. I cannot review your code.

**Reply by teleprint-me** [reply]
I did start out simply with [a regression model](https://github.com/teleprint-me/bugbear/blob/main/examples/linear.cpp). The initial link is just an MLP.

Regardless, I appreciate the reply. I eventually got the forward pass to work with the MLP.

A better line of questioning on my part might be:

- do i have to setup the input for the forward pass _in advance_ every time I construct the graph?
- can i pass in input (assuming the same dimensions) _after_ the graph has been constructed?

**Reply by slaren** [reply]
> do i have to setup the input for the forward pass in advance every time I construct the graph?
> can i pass in input (assuming the same dimensions) after the graph has been constructed?


What do you mean precisely by "setup the input" and "pass the input"? Generally:
- You can allocate inputs in the compute graph with `ggml_gallocr`, rather than preallocating them in a separate buffer, but to do so, you need to use `ggml_set_input` to flag the tensor as input
- You can load data into the input tensor at any moment before computing the graph, the data is not necessary when constructing the graph
